File: apertium-tools/simple-viewer/simple-viewer.rb
require 'libglade2'
require 'gconf2'
require 'dbus'
require 'thread'
require 'base64'

class GladeXML
  def dump_gtk_state(directory, dict)
    self.widget_names.each do |name|
      begin
      self[name].dump_state.each_pair do |key, val|
        dict[directory + "/" + name + "." + key] = val.to_s
      end
      rescue
      end
    end
  end

  def load_gtk_state(directory, dict)
    widget_props = {}
    self.widget_names.each do |name| widget_props[name] = {} end

    dict.all_entries(directory).each do |entry|
      begin
        widget_name, prop_name = entry.key[/[a-zA-Z0-9_]*[.][a-zA-Z0-9_]*/].split(".")
        widget_props[widget_name][prop_name] = entry.value
      rescue
        # Deal with badly written data
      end
    end

    widget_props.each_pair do |key, val|
      begin
        self[key].load_state(val)
      rescue
        # bummer
      end
    end
  end
end

class Gtk::Widget
  def dump_state
    {}
  end

  def load_state(p)
  end
end

class Gtk::Window
  def dump_state
    { "x_size" => self.size[0],
      "y_size" => self.size[1],
      "x_pos"  => self.position[0], 
      "y_pos"  => self.position[1] }
  end

  def load_state(p)
    self.move(p["x_pos"].to_i, p["y_pos"].to_i)
    self.resize(p["x_size"].to_i, p["y_size"].to_i)
  end
end

class Gtk::Paned
  def dump_state
    { "position" => self.position }
  end

  def load_state(p)
    self.position = p["position"].to_i
  end
end

class MainWindow
  @@PROP_DIRECTORY = "/apps/aper/simple_viewer"

  def setup_combo(info, combo)
    combo.set_model(Gtk::ListStore.new(String))
    cell = Gtk::CellRendererText.new()
    combo.pack_start(cell, true)
    combo.add_attribute(cell, "text", 0)

    @info.modes.each do |mode|
      combo.append_text(mode)
    end
    combo.set_active(0)

    combo
  end

  def options
    if @chkMarkUnknown.active?
      {"mark_unknown" => "true"}
    else
      {}
    end
  end

  def initialize(path)
    @client = GConf::Client.default
    @bus = DBus::SessionBus.instance

    @info = @bus.introspect("org.apertium.info", "/")["org.apertium.Info"]
    @mark_unknown = false

    @input_queue  = Queue.new # Thread communication queues

    @glade = GladeXML.new(path) {|handler| method(handler)}

    @buffer         = @glade["txtInput"].buffer
    @output_buffer  = @glade["txtOutput"].buffer
    @wndMain        = @glade["wndMain"]
    @dlgAbout       = @glade["dlgAbout"]
    @chkMarkUnknown = @glade["chkMarkUnknown"]

    @comboPair      = self.setup_combo(@info, @glade["comboPair"])

    @buffer.signal_connect("changed") do |w|
      @input_queue.push([self.options, w.get_text(w.start_iter, w.end_iter)])
    end

    Thread.new do
      loop do
        options, input = @input_queue.pop # Block until we have something
        options, input = @input_queue.pop until @input_queue.empty? # Then pop the queue until we have the latest item
        begin
          result = @translator.translate(options, @buffer.get_text(@buffer.start_iter, @buffer.end_iter))
        rescue Exception => e
          puts e
        end

        Gtk.idle_add do # Queue a buffer in GTK's main loop update
          @output_buffer.set_text(result)
          false
        end
      end
    end

    @glade.load_gtk_state(@@PROP_DIRECTORY, @client)
    @wndMain.visible = true
  end

  def on_btnQuit_clicked(*widget)
    @glade.dump_gtk_state(@@PROP_DIRECTORY, @client)
    Gtk.main_quit
  end

  def on_wndMain_destroy(widget)
    Gtk.main_quit
  end

  def on_wndMain_delete_event(widget, event)
    @glade.dump_gtk_state(@@PROP_DIRECTORY, @client)
    false
  end

  def on_btnAbout_clicked(widget)
    @dlgAbout.show
  end

  def on_dlgAbout_response(widget, resp)
    @dlgAbout.hide
  end

  def on_comboPair_changed(widget)
    pair_name = widget.model.get_value(widget.active_iter, 0)
    dbus_pair_name = "/" + pair_name.split("-").join("_")
    @translator = @bus.introspect("org.apertium.mode", dbus_pair_name)["org.apertium.Mode"]
    @buffer.signal_emit("changed") 
  end

  def on_chkMarkUnknown_toggled(widget)
    @buffer.signal_emit("changed") 
  end
end

Gdk::Threads.init
Gnome::Program.new('Apertium viewer', "0.2")
MainWindow.new("main_window.glade")
Gtk.main
    

File: apertium-tools/simple-viewer/simple-viewer.py
#!/usr/bin/python

import os        
import new
import sys
from Queue import Queue # A queue type to communicate between threads
import thread # A simple thread launcher
from ConfigParser import ConfigParser
# The ConfigParser is used to store preferences of the program

# GTK related imports
import gobject
import gtk
import gtk.gdk
import pygtk

import dbus # Used to connect to the Apertium D-Bus service

from i18n import _
from glade import GladeXML


class MainWindow:
    def setup_combo(self, modes, combo):
        """Set the combo box up to display simple a list of text and
        add everything in 'modes' to the list.
        """
        combo.set_model(gtk.ListStore(gobject.TYPE_STRING))
        cell = gtk.CellRendererText() # Tell the combobox we're just rendering text
        combo.pack_start(cell, True) # Add the text renderer to the combo box
        combo.add_attribute(cell, "text", 0) # Tell the combobox that we just want to show text

        for mode in modes: # Add everything in 'modes' to the combobox
            combo.append_text(mode)

        combo.set_active(0) # Set the combobox to the first entry

        return combo


    def options(self):
        if self.chkMarkUnknown.get_active():
            return {"mark_unknown": "true"}
        else:
            return {}


    def load_config(self):
        """Read the configuration data. First read default data from 'defaults.cfg'
        and then user config data from '~/.apertium-simple-viewer.cfg'. Tell the
        glade object to update stored widget properties based on the config data.
        """
        self.config.readfp(open('defaults.cfg'))
        self.config.read([os.path.expanduser('~/.apertium-simple-viewer.cfg')])
        self.glade.load_gtk_state(self.config)

    def save_config(self):
        """Store the configuration data to the user file '~/.apertium-simple-viewer.cfg'.
        Before storing, tell the glade object to dump the widget state to the configuration
        object self.config.
        """
        self.glade.dump_gtk_state(self.config)
        self.config.write(open(os.path.expanduser('~/.apertium-simple-viewer.cfg'), 'w'))
    
    def __init__(self, path):
        self.config = ConfigParser() # An object which loads/stores preferences
        self.bus = dbus.SessionBus()
        # Create a proxy object to the D-Bus object org.apertium.info/ using the interface org.apertium.Info
        self.info = dbus.Interface(self.bus.get_object("org.apertium.info", "/"), "org.apertium.Info")
        self.input_queue = Queue()  # Thread communication queue
        self.glade = GladeXML(path) # Instantiate our custom Glade class which extends the gtk.glade.GladeXML class
        self.glade.connect(self);   # Try to connect the signals defined in the glade file to methods defined in self

        self.buffer         = self.glade.get_widget("txtInput").get_buffer()
        self.output_buffer  = self.glade.get_widget("txtOutput").get_buffer()
        self.wndMain        = self.glade.get_widget("wndMain")
        self.dlgAbout       = self.glade.get_widget("dlgAbout")
        self.chkMarkUnknown = self.glade.get_widget("chkMarkUnknown")

        self.comboPair      = self.setup_combo(self.info.modes(), self.glade.get_widget("comboPair"))

        def on_changed(w):
            self.input_queue.put([self.options(), w.get_text(w.get_start_iter(), w.get_end_iter())])

        self.buffer.connect("changed", on_changed)

        thread.start_new_thread(self.translator_loop, ())

        self.load_config()
        self.wndMain.show()

        
    def translator_loop(self):
        while True:
            options, _input = self.input_queue.get()
            while not self.input_queue.empty():
                options, _input = self.input_queue.get()

            try:
                result = self.translator.translate(options, _input)

                def update_text():
                    self.output_buffer.set_text(result)
                    return False

                gobject.idle_add(update_text)

            except Exception, e:
                print e


    def on_btnQuit_clicked(self, widget):
        self.save_config()
        gtk.main_quit()

    def on_wndMain_destroy(self, widget):
        gtk.main_quit()

    def on_wndMain_delete_event(self, widget, event):
        self.save_config()
        return False

    def on_btnAbout_clicked(self, widget):
        self.dlgAbout.show()

    def on_dlgAbout_response(self, widget, response):
        self.dlgAbout.hide()

    def on_comboPair_changed(self, widget):
        pair_name = widget.get_model().get_value(widget.get_active_iter(), 0)
        dbus_pair_name = "/" + "_".join(pair_name.split("-"))
        self.translator = dbus.Interface(self.bus.get_object("org.apertium.mode", dbus_pair_name), "org.apertium.Mode")
        self.buffer.emit("changed")

    def on_chkMarkUnknown_toggled(self, widget):
        self.buffer.emit("changed")

if __name__ == "__main__":
    gtk.gdk.threads_init()
    wnd = MainWindow("main_window.glade")
    gtk.main();

